lib/Sema/TypeCheckConcurrency.cpp

```diff
@@ -5219,22 +5219,29 @@ static bool checkSendableInstanceStorage(
 
     /// Handle a stored property.
     bool operator()(VarDecl *property, Type propertyType) override {
-      // Classes with mutable properties are not Sendable.
+      // Classes with mutable properties are not Sendable unless property is
+      // actor-isolated
       if (property->supportsMutation() && isa<ClassDecl>(nominal)) {
-        if (isImplicitSendableCheck(check)) {
-          invalid = true;
-          return true;
-        }
+        ActorIsolation isolation = getActorIsolation(property);
 
-        auto behavior = SendableCheckContext(
-            dc, check).defaultDiagnosticBehavior();
-        if (behavior != DiagnosticBehavior::Ignore) {
-          property->diagnose(diag::concurrent_value_class_mutable_property,
-                             property->getName(), nominal)
-              .limitBehavior(behavior);
+        if (isolation.getKind() == ActorIsolation::Kind::Nonisolated ||
+            isolation.getKind() == ActorIsolation::Kind::Unspecified) {
+          if (isImplicitSendableCheck(check)) {
+            invalid = true;
+            return true;
+          }
+
+          auto behavior =
+              SendableCheckContext(dc, check).defaultDiagnosticBehavior();
+          if (behavior != DiagnosticBehavior::Ignore) {
+            property
+                ->diagnose(diag::concurrent_value_class_mutable_property,
+                           property->getName(), nominal)
+                .limitBehavior(behavior);
+          }
+          invalid = invalid || (behavior == DiagnosticBehavior::Unspecified);
+          return true;
         }
-        invalid = invalid || (behavior == DiagnosticBehavior::Unspecified);
-        return true;
```

Comment:
As far as I can tell, a return value of `false` in this function indicates that the property is valid for a `Sendable` conformance, so I think we should `return false;` right here instead of falling through to the `Sendable` check below. I suspect this code will fail, and think it should be valid under this change:

```swift
class NotSendable {}

final class C10: Sendable {
  @MainActor var x = NotSendable()
}
```